### ISSUE TITLE
Persisting environment variables across agent phases

### ISSUE THREAD

**Issue by vercellone** [Microsoft/vsts-tasks#3455]
Similar enough to #2111 that one could argue that issue be re-opened instead.  But, since that issue was closed before the introduction of server phases I hope this will be revisited.

**Comment by ericsciple** [comment]
Closing the issue here. Not an issue with the tasks in this repo. 

This work has already started in the [agent repo](https://github.com/Microsoft/vsts-agent) and server side.

**Comment by vercellone** [comment]
@ericsciple Glad to hear it is on the radar.  Can you provide a hyperlink to the relevant issue, PR, branch, blog, or backlog?

**Comment by ericsciple** [comment]
The backlog is internal. There's commits in the agent repo with "output variables" in the commit message.

A dependency for completing the work is moving the agent to .Net Core 1.1. We're locking down for the TFS 2017 U1 release right now, so that work is on hold for a couple weeks.

**Comment by whihathac** [comment]
really need this! :(

**Comment by ericsciple** [comment]
https://github.com/Microsoft/vsts-agent/pull/905

**Comment by whihathac** [comment]
@ericsciple - Anyone actively looking into it? Really want this as it can significantly improve our release process. Whenever we use a Server Phase like Manual intervention task, we have to re-initialize the variables in the next Agent phase.

**Comment by ericsciple** [comment]
@TingluoHuang any rough ETA on output vars?

@whihathac afaik the functionality is very close to being completed. requires server side work as well and I believe that work is already in master.

**Comment by whihathac** [comment]
Hey @ericsciple, this work is marked as complete - do we have any eta on the server side work or is it also done?

**Comment by ericsciple** [comment]
@TingluoHuang ?

**Comment by euhm** [comment]
Also curious how I can persist a variable and re-use in another phase, eg after manual intervention step.

**Comment by bryanmacfarlane** [comment]
When we're done here with output variables, they will just flow

**Comment by xenalite** [comment]
The change has been made. Is it present in Agent version 2.120.2?
I tried to set the variable with PS Inline script:
```
$variableName = 'abc'
$value = 'def'
Write-Host "##vso[task.setvariable variable=$variableName;isoutput=true;]$value"
```
It did not persist the value 'def' into the second agent phase.

**Comment by hwustrack** [comment]
Any updates on this? This would really simplify our release pipeline

**Comment by ericsciple** [comment]
https://github.com/Microsoft/vsts-agent/blob/master/docs/preview/yamlgettingstarted-phases.md#output-variables

**Comment by motowilliams** [comment]
Is the the example at https://github.com/Microsoft/vsts-agent/blob/master/docs/preview/yamlgettingstarted-phases.md#output-variables only Yaml build definitions?

Was looking into using this pattern in RM for getting output variables from an ARM deployment downstream to other phases but I can can't seem to figure out the magic incantation to solve this.

**Comment by paulsapps** [comment]
So this is not possible in "vNext" builds, only yaml?

**Comment by XmlmXmlmX** [comment]
@ericsciple @motowilliams 
> 404

**Comment by Yonnatan** [comment]
And after a year... nothing

**Comment by xenalite** [comment]
https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch#set-a-multi-job-output-variable

**Comment by XmlmXmlmX** [comment]
But:
> You cannot pass a variable from one job to another job of a build pipeline unless you use **YAML**.

**Comment by ArieHein** [comment]
And there will not be such an option in the future. As each agent is a container, enabling cross agents data persistence, means having a shared storage or mounts inside the container. The only way to do it is save the values into like a storage account / s3 and then read it as the first step of the next job.

OR you create your own containerized images with agent self registration and you run it on your k8s cluster so you have all the control you want.

**Comment by ArieHein** [comment]
The closest you can do is try to use rest api to change values of variable groups but then you loose the ability to run in parallel as the variable groups are evaluated once the agent starts.

**Comment by Yonnatan** [comment]
Thank You ,
Eventually i found a workaround by Jury rigging an Old SQL Service , to act
as sort of value holder.
